### ISSUE TITLE
Document Properties Dialog for Guis in the manual

### ISSUE THREAD

**Issue by abreubacelar** [pure-data/pddp#111]
when creating https://github.com/pd-externals/tipoftheday-plugin/issues/36 , i noticed that there is no example of properties dialog of any gui in the manual. 

There is a quick description on http://msp.ucsd.edu/Pd_documentation/x2.htm#s2.7 but it would be great if there is a better explanation and examples, since this can't go to helpfiles (there is no way of showing images of the properties dialog on the patches)

**Comment by porres** [comment]
idea is to just add this to the manual, right?

help file is very clear you can ask for properties and there is a whole section dedicated to each section of the properties and how you can also configure it with messages. I don't think that the fact we can't have a properties image in the help file is bad or troublesome...

but of course I don't oppose to adding screenshots on the manual, on the contrary. In fact I hope we can revise the manual and fill it with screenshots for all sorts of things. 

Like, why not a screenshot for gui boxes (atom/symbol/list) as well? what else?

**Comment by porres** [comment]
> i will take some screenshots and edit with gimp and try to put on the manual, I do a pr to your branch then? or htmldocs?

no idea :) it's all still confusing. I would like to have my personal html branch so I could work on it, but you know things are still unclear... I won't do it until all is settled and won't work on the manual until then...

What I also know is that IOhannes is not that happy in having to be the gatekeeper of yet another branch as he said so... 

Therefore I say we could try to avoid that one ;)

@Lucarda ?

**Comment by Lucarda** [comment]
> but of course I don't oppose to adding screenshots on the manual, on the contrary. In fact I hope we can revise the manual and fill it with screenshots for all sorts of things.

I welcome the idea but there should be certain balance.

As of how is the manual right now I don see where screen-shots of "properties dialogs" will fit or are needed. Can we discuss this or have some sort of previews? before doing any commits?

> What I also know is that IOhannes is not that happy in having to be the gatekeeper of yet another branch as he said so...

> Therefore I say we could try to avoid that one ;)

I had given all my air force planes to https://github.com/pure-data/pure-data/pull/1822 so that the manual is ordered and editable (you both know this). I think you both have to deal with me as the gatekeeper or don't count with me and be free.

Once I can get a preview on what and how @abreubacelar wants to implement this in the manual we can tell where the real commits should go (preferably not in `porres:Documentation`). The actual stuff is really a simple text edit and adding image files so is not hard at all to have previews and discussion before commits)

@abreubacelar i have given you administrator rights to https://git.nubegris.com.ar/lucarda/pd-manual-preview can you check if you can push directly? This will simplify the previews.

**Comment by porres** [comment]
> I don see where screen-shots of "properties dialogs" will fit or are needed

Why not in  http://msp.ucsd.edu/Pd_documentation/x2.htm#s2.7 ?

**Comment by Lucarda** [comment]
> Why not in http://msp.ucsd.edu/Pd_documentation/x2.htm#s2.7 ?

looks the right place. we are talking of a single screen-shot right?

**Comment by abreubacelar** [comment]
> > Why not in http://msp.ucsd.edu/Pd_documentation/x2.htm#s2.7 ?
> 
> looks the right place. we are talking of a single screen-shot right?
> 
> EDIT: do we actually need a _single_ screen-shot? why?

no, i though more about creating a new section (or a whole new appendix in a dedicated html) showing all properties and putting numbers in the image to detail what each thing does. So http://msp.ucsd.edu/Pd_documentation/x2.htm#s2.7 could have a link pointing to this section/appendix, which would be something like a "Properties Dialog Reference" or some better name...

I thought about using gimp since some properties overlap, so my plan would be:

first show one full dialog, to show a full structure, and then start from this structure top to bottom and show only small pictures where properties are different or equal for each iemgui

for example, the top layer of bang properties has "Size" while a vslider has "Width" and "Height", so i would cut only these parts of the image to show. One image showing only Size, another only for Width and Height.

And since some guis uses Size, i could group all of them below this image and explain about them, and then continue going down, for each gui and each part of the structure... i hope this doesn't sound confusing

What do you think? another option would be showing all images in full for each gui and repeating the same explanations... i don't like this one though...

**Comment by Lucarda** [comment]
> no, i though more about creating a new section ...

@abreubacelar can you start playing at https://git.nubegris.com.ar/lucarda/pd-manual-preview ?

----------------

I would be happy if we can do a dual integration. slow discussed and approved steps we can apply to `Lucarda:hugo` and a fallback to `pure-data:htmldocs`. 

I hope you understand me. Is not that I want `Lucarda:hugo` to have commits only by me. For @abreubacelar is just another target branch.

**Comment by abreubacelar** [comment]
@Lucarda no problem! i will be afk for some hours, but when i arrive home i will try to start!

**Comment by Lucarda** [comment]
Cool :)

I'll have to get some sleep for the next 10hs. If you can't push [1] to https://git.nubegris.com.ar/lucarda/pd-manual-preview fork it or download the files and then create a new repo of yours and upload the files (you already have an account).


[1] I never tried it before. In theory you should as I gave you admin rights.

**Comment by abreubacelar** [comment]
done, I will sleep now, but a basic structure is there, I will be able to continue on Friday. I will add all the screenshots and then get the info from porres helpfiles into that reference. I hope that the structure shows what I have in mind

**Comment by Lucarda** [comment]
I have an alternate branch `alt1` that shrinks `appendix A: properties dialog reference` to a single one so that there are not so many items in `index.htm`. I had also switched to no uppercase so that it follow the rest of the manual.

@abreubacelar there is a PR you can merge if you are Ok with the changes. 

@porres are we going in the right direction?

anyone wanting to preview this:

```
git clone https://git.nubegris.com.ar/lucarda/pd-manual-preview.git

git checkout alt1
```

**Comment by abreubacelar** [comment]
> I have an alternate branch `alt1` that shrinks `appendix A: properties dialog reference` to a single one so that there are not so many items in `index.htm`. I had also switched to no uppercase so that it follow the rest of the manual.

i will try to learn how to use cherry pick command to get the commit from uppercase, i didn't pay attention to this detail, thank you! 

but i don't agree with not having all items on index... this is the purpose of having an index page, right? i want it to be easy to jump to an specific part of the appendix, this will make all the appendices more useful. All Manual sections are at the top, then you go down and have all Appendices with their sections at bottom.

If someone wants to check the specific parameters for sliders, they can click on it, instead of having to enter the reference and then need to search the whole page... And thinking in the future, if someone wants to learn specifically about swap connections feature from Intelligent patching in an hypothetical appendix B they will click on its section, for example.  

Julius Orion Smith uses this approach in his online books, so for example https://ccrma.stanford.edu/~jos/filters/ has all the links there, and there are maaaany topics, i know, this index is huge... But if i need to search for something fast, i just go to that index and search for some keywords, and then jump directly to the text i want. Get to the right place in few clicks is a nice experience, i think.

So maybe, if you want to shrink we could implement this with javascript... but i fear that for some reason people don't want this on this manual right? or it is just because no one started going in that direction? 

Then maybe we could use something like accordion from bootstrap, for example? https://getbootstrap.com/docs/5.0/components/accordion/ or some arrow to collapse and expand items like this index from haskell documentation? https://downloads.haskell.org/ghc/latest/docs/libraries/ 

Or maybe we can decide this only after having all the text there, if some sections became short, maybe they could collapse into only one link. So for example we could delete 1.1.1-Width 1.1.2 Limits and etc if they are short enough, but at the index we will would have 1.1 Atom Boxes, what do you think? May we add all the info before deciding this? xD

**Comment by Lucarda** [comment]
@abreubacelar i had unshrinked `index.htm`.

I think that this part of the manual is kind of a "one time read" as in: you should not constantly refer to this section. But yes the rest of the manual is "unshrinked"

you can merge the PR and we keep the non-capitalized text.

**Comment by abreubacelar** [comment]
@Lucarda i made a middle ground commit while learning how to do cherry pick, it is merged in main https://git.nubegris.com.ar/lucarda/pd-manual-preview/commit/82362424f10061de2f74aabeeaff19f186665924

what do you think? 

i didn't merged the commit that changes the index though... but i think it is easy to do cherry pick later and merge only that commit if needed

Ok i'm very slow, i did the changes and forget to send that message, and now i pushed other commit that does the same as your pr to the index file xD sorry, do i still merge your pr? may you check the main branch now?

**Comment by Lucarda** [comment]
> do i still merge your pr?

no

**Comment by Lucarda** [comment]
> So maybe, if you want to shrink we could implement this with javascript... but i fear that for some reason people don't want this on this manual right? or it is just because no one started going in that direction?

Miller has already asked me for a non-javascript PR option in one occasion. This will be a very hard limitation if we ever include an "html object reference" as javascript is the only feasible way of providing a "search" menu.

**Comment by millerpuckette** [comment]
Sorry, I haven't been following this discussion, but it seems like I should be more explicit about my policy on introducing dependencies.  In general I'm more toleran of reasonable dependencies in the preparation of documentation, and absolutely against adding dependencies to run-time Pd (including anything needed to browse the documentation).  Certainly there shouldn't be any runnable code in HTML windows - that's just a nightmare.  I did relent and allow "css".

**Comment by Lucarda** [comment]
@millerpuckette I understand and share your view but consider:


even the simplest html file requires the browser to be in a "runnable" state.

- browser displaying a `txt` file
  - get the file and display it
  
- browser displaying an `html` file
  - get the file
  - parse it 
  - apply default css (this is a heading, this is a paragraph and this is a blue link)
    - applying a custom css file is just getting and loading the non-defaults
  - render it
    
### javascript

this can vary. 

- there are intolerable stupid things that can use lots of CPU and GPU (like constantly changing the background color of a div with a moving gradient). (these should be ruled out of course).

- there are many useful things that do not really consume resources and are not doable without javascript.


### real world measurements

https://pd.iem.sh/ implements 2 things via javascript (one may be needed if we include an "html object reference").

- theme changing (rather stupid and not really needed but without any CPU impact)

- search menu (really useful but requires CPU time)
  - this is the only feasible way of "searching" the docs without a data-base (nor in the server nor locally)
    - the search query and result page is actually processed in the browser
      - if we ever implement this there could be a warning in the page to the user and she/he can opt to not search.

    
My tests with this patch shows "audio-drops-outs" only when Pd is using around 80~90% of the CPU and *if* I do a search.

patch: [Pd-web-browser-and-javascript.zip](https://github.com/pure-data/pddp/files/10109952/Pd-web-browser-and-javascript.zip)


(only tested on an old intel i5 2.4ghz on Windows and Linux)

**Comment by porres** [comment]
I think this can be closed. We just revised chapter 2 and added new figures but I didn't find a need to add one for properties, if you want you can go ahead and add @Lucarda up to you
